File: ProductOfArray.java
import java.util.Arrays;

public class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] prefix = new int[n];
        int res = 1;
        prefix[0] = nums[0];

        for (int i = 1; i < n; i++) {
            prefix[i] = prefix[i - 1] * nums[i];
        }

        for (int i = n - 1; i >= 1; i--) {
            prefix[i] = prefix[i - 1] * res;
            res *= nums[i];
        }
        prefix[0] = res;
        return prefix;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = {1, 2, 3, 4};
        int[] result = solution.productExceptSelf(nums);
        System.out.println(Arrays.toString(result));
    }
}

File: ProductOfArray.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
         int n = nums.size();
         vector<int>prefix(n,1);
         int res = 1;
         prefix[0]=nums[0];

         for(int i=1;i<n;i++){
               prefix[i] = prefix[i-1]*nums[i];
         }

         for(int i=n-1;i>=1;i--){
               prefix[i] = prefix[i-1]*res;
               res *= nums[i];
         }
         prefix[0] = res;
         return prefix;
    }
};
